src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al

```diff
@@ -0,0 +1,518 @@
+// Licensed under the MIT License. See License.txt in the project root for license information.
+// ------------------------------------------------------------------------------------------------
+
+namespace Microsoft.Foundation.NoSeries;
+
+using System.Telemetry;
+using System.AI;
+using System.Text.Json;
+
+codeunit 324 "No. Series Copilot Impl."
+{
+    var
+        IncorrectCompletionErr: Label 'Incorrect completion. The property %1 is empty', Comment = '%1 = property name';
+        IncorrectCompletionNumberOfGeneratedNoSeriesErr: Label 'Incorrect completion. The number of generated number series is incorrect. Expected %1, but got %2', Comment = '%1 = Expected Number, %2 = Actual Number';
+        TextLengthIsOverMaxLimitErr: Label 'The property %1 exceeds the maximum length of %2', Comment = '%1 = property name, %2 = maximum length';
+        DateSpecificPlaceholderLbl: Label '{current_date}', Locked = true;
+        NotAbleToGenerateNumberSeriesTryToRephraseErr: Label 'Sorry, I am not able to generate the number series. Try to rephrase your request or provide more details.';
+
+    internal procedure GetNoSeriesSuggestions()
+    var
+        FeatureTelemetry: Codeunit "Feature Telemetry";
+        NoSeriesCopilotRegister: Codeunit "No. Series Copilot Register";
+        NoSeriesCopilotImpl: Codeunit "No. Series Copilot Impl.";
+        AzureOpenAI: Codeunit "Azure OpenAI";
+        NoSeriesCopilot: Page "No. Series Proposal";
+    begin
+        NoSeriesCopilotRegister.RegisterCapability();
+        if not AzureOpenAI.IsEnabled(Enum::"Copilot Capability"::"No. Series Copilot") then
+            exit;
+
+        FeatureTelemetry.LogUptake('0000LF4', NoSeriesCopilotImpl.FeatureName(), Enum::"Feature Uptake Status"::Discovered); //TODO: Update signal id
+
+        NoSeriesCopilot.LookupMode := true;
+        NoSeriesCopilot.Run();
+    end;
+
+    procedure Generate(var NoSeriesProposal: Record "No. Series Proposal"; var ResponseText: Text; var NoSeriesGenerated: Record "No. Series Proposal Line"; InputText: Text)
+    var
+        TokenCountImpl: Codeunit "AOAI Token";
+        SystemPromptTxt: SecretText;
+        CompletePromptTokenCount: Integer;
+        Completion: Text;
+    begin
+        Clear(ResponseText);
+        SystemPromptTxt := GetToolsSystemPrompt();
+
+        CompletePromptTokenCount := TokenCountImpl.GetGPT4TokenCount(SystemPromptTxt) + TokenCountImpl.GetGPT4TokenCount(InputText);
+        if CompletePromptTokenCount <= MaxInputTokens() then begin
+            Completion := GenerateNoSeries(SystemPromptTxt, InputText);
+            if CheckIfValidCompletion(Completion) then begin
+                SaveGenerationHistory(NoSeriesProposal, InputText);
+                CreateNoSeries(NoSeriesProposal, NoSeriesGenerated, Completion);
+            end else
+                ResponseText := Completion;
+        end;
+    end;
+
+    procedure ApplyProposedNoSeries(var NoSeriesGenerated: Record "No. Series Proposal Line")
```

Comment:
same, i'd like the variable name to somewhat match the object name. Variables can add descriptors to make it more precise but the more they differ the more I think we simply need better names...